lib/Dialect/ESI/Passes/ESILowerPorts.cpp

```diff
@@ -39,273 +34,30 @@ inline static StringRef getStringAttributeOr(Operation *op, StringRef attrName,
-          origPort.loc, data);
-
-    for (size_t idx = 0, e = elements.size(); idx < e; ++idx) {
-      auto field = elements[idx];
-      Value fieldValue = explode ? explode->getResult(idx) : Value();
-      rewriter.createNewOutput(origPort, "_" + field.name.getValue(),
-                               field.type, fieldValue, dataPorts[idx]);
-    }
-  }
-}
-
-// Map the data ports coming off an instance back into the original ports. If
-// the original type was a channel of a struct (and we've been asked to flatten
-// it), construct the original struct from the new ports.
-Value SignalingStandard::mapOutputDataPorts(OpBuilder &b,
-                                            ArrayRef<Backedge> newResults) {
-  auto chanTy = origPort.type.dyn_cast<ChannelType>();
-  Type dataPortType = chanTy ? chanTy.getInner() : origPort.type;
-  hw::StructType origStruct =
-      chanTy ? dataPortType.dyn_cast<hw::StructType>() : hw::StructType();
-
-  if (!rewriter.shouldFlattenStructs() || !origStruct)
-    return newResults[dataPorts[0].argNum];
-
-  SmallVector<Value, 16> fieldValues;
-  for (auto portInfo : dataPorts)
-    fieldValues.push_back(newResults[portInfo.argNum]);
-  return b.create<hw::StructCreateOp>(origPort.loc, origStruct, fieldValues);
-}
-
-/// We consider non-ESI ports to be ad-hoc signaling or 'raw wires'. (Which
-/// counts as a signaling protocol if one squints pretty hard). We mostly do
-/// this since it allows us a more consistent internal API.
-class RawWires : public SignalingStandard {
-public:
-  using SignalingStandard::SignalingStandard;
-
-  void mapInputSignals(OpBuilder &b, Operation *inst, Value instValue,
-                       SmallVectorImpl<Value> &newOperands,
-                       ArrayRef<Backedge> newResults) override {
-    mapInputDataPorts(b, instValue, newOperands);
-  }
-  void mapOutputSignals(OpBuilder &b, Operation *inst, Value instValue,
-                        SmallVectorImpl<Value> &newOperands,
-                        ArrayRef<Backedge> newResults) override {
-    instValue.replaceAllUsesWith(mapOutputDataPorts(b, newResults));
-  }
-
-private:
-  void buildInputSignals() override {
-    Value newValue = buildInputDataPorts();
-    if (body)
-      body->getArgument(origPort.argNum).replaceAllUsesWith(newValue);
-  }
-
-  void buildOutputSignals() override {
-    Value output;
-    if (body)
-      output = body->getTerminator()->getOperand(origPort.argNum);
-    buildOutputDataPorts(output);
-  }
 };
 
 /// Implement the Valid/Ready signaling standard.
-class ValidReady : public SignalingStandard {
+class ValidReady : public ESISignalingStandad {
 public:
-  using SignalingStandard::SignalingStandard;
+  ValidReady(PortConverterImpl &converter, hw::PortInfo origPort)
+      : ESISignalingStandad(converter, origPort), validPort(origPort),
+        readyPort(origPort),
+        validSuffix(getStringAttributeOr(converter.getModule(),
```

Comment:
Since we're not storing these strings in `ChannelRewriter` (which allowed us to amortize their storage over all the ports), we don't want to store them here since they're only used once -- to create the `PortInfo`s. So don't store them at all, just call them when creating the ports. It was just a minor optimization to begin with.